### PR TITLE
Throw PSR-18 exceptions

### PR DESCRIPTION
**Pull request by gmponos**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | yes
| Deprecations?   | no
| Related tickets | #61
| License         | MIT

#### What's in this PR?

Explained on #61.

#### Why?

Be compatible with PSR-18

#### Checklist

- [x] Updated CHANGELOG.md to describe BC breaks / deprecations | new feature | bugfix

#### To Do

- [x] ~~This is a BC but still if the package needs to be PSR-18 compliant it needs to be done.~~

### CODE REVIEW

**Comment by sagikazarmark** [comment]
Well, I don't think it's a BC break for two reasons:

1. It wasn't compliant with the standard, so it qualifies as a bug fix.
2. We can actually do it in a BC way, by extending `RuntimeException` and `UnexpectedValueException` respectively.

The standard specifies interfaces, so I think we can do that.

**Comment by gmponos** [comment]
OK.. so I guess the current state of this PR is not fine?

We need to go [here](https://github.com/php-http/httplug/tree/master/src/Exception) and create these exceptions or do you want them to be created in the current repo?

**Comment by sagikazarmark** [comment]
I'm failing to see why we need to create these exceptions there. Why can't we create them in this repo? I could even live with anonymous classes since we are on PHP 7.x now.

**Comment by sagikazarmark** [comment]
> No need to bother with a the other exception since TransferException already extends \RuntimeException.

Does it actually fit the TransferException. As far as I can see it's just a catchall for exceptions.

(Note: We could check throwable there as well?)

**Comment by gmponos** [comment]
> (Note: We could check throwable there as well?)

👍 nice one... Done.

> I'm failing to see why we need to create these exceptions there. Why can't we create them in this repo? 

I am just asking in order to have clear directions ✌️ 

> Does it actually fit the TransferException. 

You guys wrote the PSR.. but the way I understand it from the examples of the exceptions at the [docblocks](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-18-http-client.md#requestexceptioninterface) it might be a TransferException or it might even be a `RequestException` since Guzzle Handlers [throws typical Exceptions](https://github.com/guzzle/guzzle/blob/master/src/Handler/CurlFactory.php#L370) when they fail to Execute the request for Runtime reasons, (like when a certificate does not exists).

Not sure about this. I will do as you think, but if you go with not throwing a `TransferException` please provide me with a proper name because I am out of ideas on this.

> write a change log

I was not expecting to write a changelog. I hope it is good. Until we think of a name maybe you could have a look there.

**Comment by gmponos** [comment]
WIP because we must wait for #63 to be merged first.

**Comment by dbu** [comment]
i merged #63, please rebase (and solve the conflicts)

File: src/Promise.php [hunk 1]
```diff
@@ -62,9 +63,9 @@ public function __construct(PromiseInterface $promise, RequestInterface $request
             } elseif ($reason instanceof GuzzleExceptions\GuzzleException) {
                 $this->exception = $this->handleException($reason, $request);
             } elseif ($reason instanceof \Throwable) {
-                $this->exception = new \RuntimeException('Invalid exception returned from Guzzle6', 0, $reason);
+                $this->exception = new HttplugException\TransferException('Invalid exception returned from Guzzle6', 0, $reason);
             } else {
-                $this->exception = new \UnexpectedValueException('Reason returned from Guzzle6 must be an Exception');
+                $this->exception = new UnexpectedValueException('Reason returned from Guzzle6 must be an Exception');
```
[the remark this answers is not in the document]
Comment (gmponos) [reply src/Promise.php]: 🤔 

`use GuzzleHttp\Exception as GuzzleExceptions;`

Under `GuzzleHttp\Exception` there is no `UnexpectedValueException`

[the remark this answers is not in the document]
Comment (gmponos) [reply src/Promise.php]: Unless if you meant `HttplugException\UnexpectedValueException`?

**Comment by gmponos** [comment]
> do you want to adjust the test case you refactored in #63 to ensure that all exceptions thrown implement httplug exception interfaces too?

👍 .. it's been a long time and I totally forgot about it.

File: src/Promise.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (gmponos) [reply src/Promise.php]: Also `HttplugException\` does not have `UnexpectedValueException`

unless if you want me to do something like this: 

```
use Http\Adapter\Guzzle6\Exception as Guzzle6AdapterException;
...
$this->exception = new Guzzle6AdapterException\UnexpectedValueException('Reason returned from Guzzle6 must be an Exception');
```

**Comment by dbu** [comment]
i released 2.0.1 with this fix. thanks again!

**Pull request merged**